ocaml/idl/markdown_backend.ml

```diff
@@ -0,0 +1,373 @@
+
+(** Add namespaces (separated by _) to each field name *)
+let flatten stuff =
+  let rec f ns = function
+    | Field fr -> Field { fr with field_name = ns ^ fr.field_name}
+    | Namespace(ns', contents) -> Namespace("", List.map (f (ns ^ ns' ^ "_")) contents)
+  in
+  f "" stuff
+
+let string_of_qualifier = function
+  | StaticRO   -> "_RO/constructor_"
+  | DynamicRO  -> "_RO/runtime_"
+  | RW         -> "_RW_"
+
+
+let string_of_open_product release =
+  if release.internal_deprecated_since = None then "" else "**Deprecated.** "
+
+(* Make a markdown section for an API-specified message *)
+let markdown_section_of_message printer x =
+  let return_type = of_ty_opt_verbatim x.msg_result in
+
+  printer (sprintf "#### RPC name: %s" (escape x.msg_name));
+  printer "";
+  if x.msg_release.internal_deprecated_since <> None then
+  begin
+    printer "This message is deprecated.";
+    printer ""
+  end;
+  printer "_Overview:_";
+  printer "";
+  printer (escape x.msg_doc);
+  printer "";
+  printer "_Signature:_";
+  printer "";
+  printer "```";
+  printer (of_ty_opt_verbatim x.msg_result);
+  printer (sprintf "%s (%s)"
+    x.msg_name
+    (String.concat ", "
+      ((if x.msg_session then "session_id s" else "")::
+        (List.map (fun p -> of_ty_verbatim p.param_type ^ " " ^ p.param_name) x.msg_params)))
+  );
+  printer "```";
+
+  if x.msg_params <> [] then begin
+    printer "_Arguments:_";
+    printer "";
+    printer "|type|name|description|";
+    printer "|:---|:---|:---|";
+
+    let get_param_row p = sprintf "|`%s`|%s|%s|" (of_ty_verbatim p.param_type) (escape p.param_name) (escape p.param_doc) in
+    List.iter (fun p -> printer (get_param_row p)) x.msg_params;
+    printer "";
+
+    printer (sprintf "_Return Type:_ `%s`" return_type);
+    printer "";
+    let descr= desc_of_ty_opt x.msg_result in
+    if descr <> ""  then
+      (printer (escape descr);
+      printer "")
+  end;
+
+  if x.msg_errors <> [] then begin
+    let error_codes = List.map (fun err -> sprintf "`%s`" err.err_name) x.msg_errors in
+    printer (sprintf "_Possible Error Codes:_ %s"
+                   (String.concat ", " error_codes));
+    printer "";
+  end
+
+let print_field_table_of_obj printer x =
+  printer (sprintf "### Fields for class: "^(escape x.name));
+  printer "";
+  if x.contents=[] then
+    printer ("Class "^(escape x.name)^" has no fields.")
+  else begin
+    printer "|Field|Type|Qualifier|Description|";
+    printer "|:---|:---|:---|:---|";
+
+    let rec print_field_content printer = function
+    | Field{release; qualifier; field_name=name; ty; field_description=description} ->
+        printer (sprintf "|%s|`%s`|%s|%s%s|"
+          (escape name) (of_ty_verbatim ty) (string_of_qualifier qualifier)
+          (string_of_open_product release) (escape description))
+    | Namespace(_, fields) -> List.iter (print_field_content printer) fields
+    in
+
+    List.iter (fun x -> print_field_content printer (flatten x)) x.contents
```

Comment:
As above, I think we could simplify this to somethig like `x |> Datamodel_utils.fields_of_obj |> List.map Datamodel_utils.wire_name_of_field |> List.iter print_field_content` using the helpers in `Datamodel_utils`. Then `print_field_content` would only have to deal with `Field`, and could be non-recursive (as the recursion on Namespaces is already done by `Datamodel_utils.fields_of_obj`.